ksql-engine/src/main/java/io/confluent/ksql/topic/DefaultTopicInjector.java

```diff
@@ -0,0 +1,153 @@
+  }
+
+  DefaultTopicInjector(
+      final KafkaTopicClient topicClient,
+      final MetaStore metaStore) {
+    this.topicClient = topicClient;
+    this.metaStore = metaStore;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> PreparedStatement<T> forStatement(
+      final PreparedStatement<T> statement,
+      final KsqlConfig ksqlConfig,
+      final Map<String, Object> propertyOverrides) {
+    if (!(statement.getStatement() instanceof CreateAsSelect)) {
+      return statement;
+    }
+
+    final CreateAsSelect cas = (CreateAsSelect) statement.getStatement();
+
+    final Analysis analysis = new QueryAnalyzer(
+        metaStore,
+        ksqlConfig.getString(KsqlConfig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG))
+        .analyze(SqlFormatter.formatSql(cas), cas.getQuery(), Optional.of(cas.getSink()));
+
+    final String topic = topicName(cas, ksqlConfig);
+    final TopicDescription description = describeSource(analysis, topicClient);
+    final int partitions = numPartitions(cas, ksqlConfig, propertyOverrides, description);
+    final short replicas = numReplicas(cas, ksqlConfig, propertyOverrides, description);
+
+    final Map<String, Expression> properties = new HashMap<>(cas.getProperties());
+    properties.put(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY, new StringLiteral(topic));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_REPLICAS, new IntegerLiteral(replicas));
+    properties.put(KsqlConstants.SINK_NUMBER_OF_PARTITIONS, new IntegerLiteral(partitions));
+
+    final CreateAsSelect withTopic = cas.copyWith(properties);
+    final String withTopicText = SqlFormatter.formatSql(withTopic) + ";";
+
+    return (PreparedStatement<T>) PreparedStatement.of(withTopicText, withTopic);
+  }
+
+  private static TopicDescription describeSource(
+      final Analysis analysis,
+      final KafkaTopicClient topicClient
+  ) {
+    final StructuredDataSource theSource = analysis.getTheSource();
+    final String kafkaTopicName = theSource.getKsqlTopic().getKafkaTopicName();
+    return topicClient.describeTopic(kafkaTopicName);
+  }
+
+  private static String topicName(
+      final CreateAsSelect cas,
+      final KsqlConfig cfg
+  ) {
+    final Expression topicProperty = cas.getProperties().get(DdlConfig.KAFKA_TOPIC_NAME_PROPERTY);
+    return (topicProperty == null)
+        ? cfg.getString(KsqlConfig.KSQL_OUTPUT_TOPIC_NAME_PREFIX_CONFIG) + cas.getName().getSuffix()
+        : StringUtils.strip(topicProperty.toString(), "'");
+  }
+
+  private static int numPartitions(
+      final CreateAsSelect cas,
+      final KsqlConfig cfg,
+      final Map<String, Object> propertyOverrides,
+      final TopicDescription description
+  ) {
+    final Expression partitions = cas.getProperties().get(KsqlConstants.SINK_NUMBER_OF_PARTITIONS);
+    if (partitions != null) {
+      return Integer.parseInt(partitions.toString());
```

Comment:
> a better way (IMHO) is to look at the type of the expression

Sure, this is reasonable. I think it's a little bit out of scope, especially because it's rather complicated unless properties returns `Map<String, Literal>` (which is not a minor change). Until then, I'll put in some actionable error messages if parse fails and I'll keep `toString`.